src/logic/player.cc

```diff
@@ -1479,14 +1484,52 @@ void Player::update_vision(const Area<FCoords>& area, bool force_visible) {
 	if (egbase().objects().is_cleaning_up()) {
 		return;
 	}
+
+	// Build a list of nearby objects.
+	// This way we only evaluate all objects (which is very expensive) once.
+	// When updating individual fields, only the objects from this list are evaluated.
+	std::list<std::pair<int, const MapObject*>> nearby_objects;
+	for (const PlayerNumber& p : team_player_) {
+		Player& player = *egbase().get_player(p);
+		for (const MapObject* seer : player.seers_) {
+			if (seer->descr().type() >= MapObjectType::BUILDING) {
+				assert(is_a(Building, seer));
+				const Building* b = static_cast<const Building*>(seer);
+				if (b->is_seeing()) {
+					int dist =
+					   egbase().map().calc_distance(area, b->get_position()) - b->descr().vision_range();
+					if (dist <= area.radius) {
+						nearby_objects.push_back(std::make_pair(dist, b));
+					}
+				}
+			} else {
+				// currently only buildings and bobs can see fields
+				assert(is_a(Bob, seer));
+				const Bob* b = static_cast<const Bob*>(seer);
+				int dist =
+				   egbase().map().calc_distance(area, b->get_position()) - b->descr().vision_range();
+				if (dist <= area.radius) {
+					nearby_objects.push_back(std::make_pair(dist, b));
+				}
+			}
+		}
+	}
+	// Sort the list, so the objects that are more likely to be a hit are evaluated first.
```

Comment:
How much is the performance gain by sorting them, vs the additional complexity introduced by performing the sorting?

Reply:
I didn't keep the numbers but AFAIR the gain was negligible in the average case, but substantial in the dismantling tower test. Basically it benefits big area seers, should also help with scouts (the tested savegame didn't have them) etc.

Reply:
Dismantling is 3-5 times slower without sorting.